core/trino-main/src/main/java/io/trino/operator/aggregation/DecimalAverageAggregation.java

```diff
@@ -161,28 +161,31 @@ public static void inputLongDecimal(Type type, LongDecimalWithOverflowAndLongSta
 
     private static void accumulateValueInState(Slice unscaledValue, LongDecimalWithOverflowAndLongState state)
     {
-        accumulateAndUpdateOverflow(unscaledValue, state);
-        state.setLong(state.getLong() + 1);
-    }
-
-    private static void initializeIfNeeded(LongDecimalWithOverflowAndLongState state)
-    {
-        if (state.getLongDecimal() == null) {
-            state.setLongDecimal(UnscaledDecimal128Arithmetic.unscaledDecimal());
+        Slice sum = state.getLongDecimal();
+        if (sum == null) {
+            sum = UnscaledDecimal128Arithmetic.unscaledDecimal();
+            state.setLongDecimal(sum);
         }
+        long overflow = UnscaledDecimal128Arithmetic.addWithOverflow(sum, unscaledValue, sum);
+        state.addOverflow(overflow);
+        state.addLong(1);
     }
 
     public static void combine(LongDecimalWithOverflowAndLongState state, LongDecimalWithOverflowAndLongState otherState)
     {
-        state.setLong(state.getLong() + otherState.getLong());
-        state.setOverflow(state.getOverflow() + otherState.getOverflow());
+        state.addLong(otherState.getLong());
```

Comment:
it would be nice to have some comments what `addLong` means (e.g updating row counter)